model-optimizer/extensions/ops/sparse_reshape.py

```diff
@@ -28,26 +29,75 @@ def supported_attrs(self):
+            elif x != -1:
+                num_of_output_elements *= x
+
+        # input_shape = [dyn, 5, 6], new_shape = [0, -1] => output_shape [dyn, 30]
+        dyns_are_copied = True  # true if there are no dyns or all of them are copied with "0" magic value
+        if not is_fully_defined(input_shape):
+            for index, x in enumerate(input_shape):
+                if x is dynamic_dimension:
+                    if index >= len(new_shape) or new_shape[index] != 0:
+                        dyns_are_copied = False
+
+        undefined_dim = dynamic_dimension
+        if num_of_output_elements is not dynamic_dimension and dyns_are_copied and \
+                is_fully_defined(new_shape):
+            undefined_dim = num_of_input_elements // num_of_output_elements
+        output_shape = []
+        for index, x in enumerate(new_shape):
+            if x == 0 and node.has_and_set('special_zero'):
+                output_shape.append(input_shape[index])
+            elif x == -1:
+                output_shape.append(undefined_dim)
+            else:
+                output_shape.append(x)
+
+        # even if the new_shape contains some dynamic values we can calculate the actual value by deducing it from the
+        # input shape if it is static: input_shape = [5, 3, 8], new_shape = [4, d] => output_shape = [4, 30]
+        if is_fully_defined(input_shape) and not is_fully_defined(new_shape):
+            dynamic_indices = np.argwhere([item is dynamic_dimension for item in new_shape])
+            num_of_output_elements = 1
+            if dynamic_indices.size == 1:
+                for index, x in enumerate(new_shape):
+                    if x == 0 and node.has_and_set('special_zero'):
+                        num_of_output_elements *= input_shape[index]
+                    elif x is not dynamic_dimension and x != -1:
+                        num_of_output_elements *= x
+            assert num_of_input_elements % num_of_output_elements == 0, \
+                'Incorrect number of output elements deduced for node {}: '.format(name)
+            output_shape[dynamic_indices[0][0]] = num_of_input_elements // num_of_output_elements
+
+        assert not is_fully_defined(input_shape) or not is_fully_defined(output_shape) or \
+               np.prod(input_shape) == np.prod(output_shape), \
+            "Number of elements in input {} and output {} of dynamic reshape node {} mismatch" \
+            "".format(input_shape, output_shape, name)
+
+        node.out_port(1).data.set_value(shape_array(output_shape))
         output_indices_shape = np.concatenate((input_indices_shape[0:1], new_shape_shape))
         node.out_port(0).data.set_shape(output_indices_shape)
 
         # TODO: implement constant value propagation for common case
-        if np.array_equal(input_shape_value, output_shape_value) and input_indices_value is not None:
+        if np.array_equal(input_shape, output_shape) and input_indices_value is not None:
```

Comment:
np.array_equal works incorrectly for masked array. Consider using strict_compare_tensors() or compatible_shapes().

Reply:
Thanks for noting. Agreed and replaced with `strict_compare_tensors`